### PR TITLE
docs(how-to-guides): add traffic_light_ssd_fine_detector node neural network training documentation

### PR DESCRIPTION
**Pull request by Zeysthingz**
## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The Reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by scepter914** [comment]
@Mingyu1991 cc. @miursh 

This PR is document maintenance related to machine learning module for traffic light recognition.
Could you please review this PR?

**Comment by Mingyu1991** [comment]
LGTM.

**Review by Mingyu1991 — APPROVED** [review]
LGTM

File: docs/how-to-guides/ssd_fine_detector_network_train.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# How to Train SSD Mobilenetv2 Lite Inside traffic_light_ssd_fine_detector Node
+
+SSD MobilenetV2 Lite is a fine detector for traffic light recognition. This node is a part of the traffic light recognition and classification pipeline. You can download original repo and pretrained weights from here <https://github.com/qfgaohao/pytorch-ssd>. Output of this node is an array of traffic light bounding boxes and their labels and `traffic_light_classifier` node uses this output to classify the traffic lights.
+
+Neural Network supports two different annotation format as Open-Images and PASCAL-VOC datasets. We can train our dataset after editing label format according to these formats.
+With original repo requirements we are not able to train the neural network so requirements should be updated as below.
```
Comment (kaancolak) [inline docs/how-to-guides/ssd_fine_detector_network_train.md:6]:
```suggestion
SSD MobilenetV2 Lite is a fine detector for traffic light recognition. This node is a part of the traffic light recognition and classification pipeline. You can download the original repo and pretrained weights from here <https://github.com/qfgaohao/pytorch-ssd>. The output of this node is an array of traffic light bounding boxes and their labels and `traffic_light_classifier` node uses this output to classify the traffic lights.

Neural Network supports two different annotation formats as Open-Images and PASCAL-VOC datasets. We can train our dataset after editing the label format according to these formats.
With original repo requirements, we are not able to train the neural network so requirements should be updated as below.
```

File: docs/how-to-guides/ssd_fine_detector_network_train.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+- `Caffe2`
+- `torchvision==0.13.0+cu116`
+- `torch==1.12.0+cu116`
+- `onnx==1.8.0`
+- `onnxruntime==1.12.1`
+
+## How to Train Model
+
+If we are going to train our custom dataset in open images dataset format first we need divide dataset as train test and validation subsets and put them in a folder named as `open_images` . Also there should be train.csv test.cvs and validation.csv inside open-images folder.
+After this step we should change label names to `BACKGROUND` and `traffic_light` in models/open-images-model-labels.txt file. Then we can start training by running following command:
```
Comment (kaancolak) [inline docs/how-to-guides/ssd_fine_detector_network_train.md:20]:
```suggestion
If we are going to train our custom dataset in open images dataset format first we need to divide the dataset as train test and validation subsets and put them in a folder named `open_images`. Also, there should be train.csv test.cvs and validation.csv inside the open-images folder.
After this step, we should change label names to `BACKGROUND` and `traffic_light` in models/open-images-model-labels.txt file. Then we can start training by running the following command:
```

File: docs/how-to-guides/ssd_fine_detector_network_train.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+- `--validation_epochs` : Number of epochs between validation
+- `--num_epochs` : Number of epochs to train
+- `--base_net_lr` : Learning rate for base network
+- `--batch_size` : Batch size. Batch size should be multiple of 2
+
+For all parameters you can check `train_ssd.py` file.
+
+## How to Evaluate Model
+
+To evaluate the model run following command:
```
Comment (kaancolak) [inline docs/how-to-guides/ssd_fine_detector_network_train.md:45]:
```suggestion
To evaluate the model run the following command:
```

File: docs/how-to-guides/ssd_fine_detector_network_train.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+```bash
+python eval_ssd.py --net mb2-ssd-lite ~/open-images/test --trained_model models/mb2-ssd-lite-mp-0_686.pth --label_file models/open-images-model-labels.txt
+```
+
+## How to Convert .pth to ONNX
+
+To convert the model to ONNX run following command:
```
Comment (kaancolak) [inline docs/how-to-guides/ssd_fine_detector_network_train.md:53]:
```suggestion
To convert the model to ONNX run the following command:
```

File: docs/how-to-guides/ssd_fine_detector_network_train.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+```bash
+ python convert_to_caffe2_models.py mb2-ssd-lite models/mb2-ssd-lite-mp-0_686.pth models/open-images-model-labels.txt
+```
+
+## How to Run Model Inference on Image
+
+To run the example run following command:
```
Comment (kaancolak) [inline docs/how-to-guides/ssd_fine_detector_network_train.md:61]:
```suggestion
To run the example run the following command:
```

**Review by kaancolak — APPROVED** [review]
LGTM

**Comment by yukkysaito** [comment]
Thank you for PR.

The document looks good, but I would like to discuss placement.
My understanding is that the autoware-document is placed in the implementation-independent documentation, and the implementation-dependent documentation is placed in the package.
Since this document depends on traffic_light_ssd_fine_detector, it might be better to have it in the package.
How about posting a link from the how to guide page to the introduction of the packages that depend on machine learning?

- https://github.com/autowarefoundation/autoware.universe/tree/main/perception/traffic_light_ssd_fine_detector
- https://autowarefoundation.github.io/autoware.universe/main/perception/traffic_light_ssd_fine_detector/

**Comment by mitsudome-r** [comment]
@yukkysaito Good point, but I think it would be user friendly if we have a page for neural network training in Autoare Documentation as well since that's where most of the users would read first. How about we move the content to the package README and create a page in autoware-documentation that collects the link to training documentation guide for each package?

**Comment by Zeysthingz** [comment]
> @yukkysaito Good point, but I think it would be user friendly if we have a page for neural network training in Autoare Documentation as well since that's where most of the users would read first. How about we move the content to the package README and create a page in autoware-documentation that collects the link to training documentation guide for each package?

Thank you for your comment, that also can be also  usefull and more organised. I can work on it if it's okey ?

**Comment by yukkysaito** [comment]
>Thank you for your comment, that also can be usefull and more organised. I can work on it if it's okey ?

Sorry for late reply.
Sure. Thank you for your cooperation. :+1:
